service/history/timerQueueProcessor.go

```diff
@@ -217,32 +221,33 @@ continueProcessor:
 				t.logger.Debug("Timer queue processor pump shutting down.")
 				return nil
 
-			case <-timerChan:
+			case <-timerGate.FireChan():
 				// Timer Fired.
 
-			case newEarlistTime = <-t.newTimerCh:
-				// New Timer has arrived.
-
 			case <-updateAckChan:
 				t.ackMgr.updateAckLevel()
 				continue continueProcessor
-			}
-		}
 
-		if !newEarlistTime.IsZero() {
-			t.metricsClient.IncCounter(metrics.TimerQueueProcessorScope, metrics.NewTimerNotifyCounter)
-			t.logger.Debugf("Woke up by the timer")
+			case <-t.newTimerCh:
+				t.newTimeLock.Lock()
+				newTime := t.newTime
+				t.newTime = time.Time{}
```

Comment:
can we define a null time const and set it to null time instead of creating a new struct each time.

Reply:
const can only be character, string, boolean, or numeric values.

i can change the variable type to *time.Time